### PR TITLE
Rya 234 statement metadata optimizer

### PR DESCRIPTION
**Pull request by meiercaleb**
Added a package so that reified queries can be issued to Rya in an efficient way.  Using the StatementMetada field of a RyaStatement, users can issue reified queries that includes contextual properties of each triple.  Reified queries are matched with a StatementMetadataNode, which is inserted
into the query plan using StatementMetadataOpitimizer.

### Tests
Tests have been included to test the functionality of the StatementMetadataNode, StatementMetadataOptimizer and related classes for both Mongo backed Rya and Accumulo backed Rya.

### Links
[Jira](https://issues.apache.org/jira/browse/RYA-234)

### Checklist
- [ ] Code Review
- [x] Squash Commits

#### People To Reivew
Andrew Smith, Puja Valiyil

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -61,7 +61,15 @@ Run the command to build the code `mvn clean install`
...
-Note: To perform a build of the geomesa/lucene indexing, run the build with the profile 'indexing' `mvn clean install -P indexing`
+Note: The following profiles are available to tailor the build:
+
+
+| Profile ID | Purpose |
+| ---------- | ------- |
+| geoindexing | perform a build of the geomesa/lucene indexing |
+| mongodb | build with mongoDB configuration (defaults to accumulo) |
+
+To run the build with the profile 'geoindexing' `mvn clean install -P geoindexing`.
```
Comment (pujav65) [inline README.md]: was this for another jira ticket.  I'm confused over why it was bundled here or if this is just a bad rebase or something

File: common/rya.api/src/main/java/org/apache/rya/api/domain/StatementMetadata.java [hunk 2]
```diff
@@ -19,36 +19,50 @@
...
-    }
-    
-    public void addMetadata(String key, String value){
-        metadataMap.put(key, value);
+        gson = new GsonBuilder().enableComplexMapKeySerialization()
+                .registerTypeHierarchyAdapter(RyaType.class, new RyaTypeAdapter()).create();
     }
 
     public StatementMetadata(byte[] value) throws RdfDAOException {
+        super();
```
Comment (pujav65) [inline common/rya.api/src/main/java/org/apache/rya/api/domain/StatementMetadata.java]: whats the super class here?

File: common/rya.api/src/main/java/org/apache/rya/api/domain/StatementMetadata.java [hunk 3]
```diff
@@ -19,36 +19,50 @@
...
+    @SuppressWarnings("serial")
+    private Type type = new TypeToken<Map<RyaURI, RyaType>>(){}.getType();
+
     public StatementMetadata() {
-        
-    }
-    
-    public void addMetadata(String key, String value){
-        metadataMap.put(key, value);
+        gson = new GsonBuilder().enableComplexMapKeySerialization()
```
Comment (pujav65) [inline common/rya.api/src/main/java/org/apache/rya/api/domain/StatementMetadata.java]: should we be resetting this static object repeatedly?  Why not just have a static initialization?

File: README.md (hunk 1 above)
Comment (meiercaleb) [inline README.md]: By product of the rebase.  Initially this branch was clone off of a branch that Andrew was working on (he had made some progress on the external optimizer framework that I picked up).  That pulled in a number of other commits that have been merged -- I didn't know whether you wanted my PR to highlight only my changes or if you wanted other peoples commits included.

File: common/rya.api/src/main/java/org/apache/rya/api/domain/StatementMetadata.java (hunk 2 above)
Comment (meiercaleb) [inline common/rya.api/src/main/java/org/apache/rya/api/domain/StatementMetadata.java]: Whoops!  Good catch.  I think I meant to put this().

File: common/rya.api/src/main/java/org/apache/rya/api/domain/StatementMetadata.java (hunk 3 above)
Comment (meiercaleb) [inline common/rya.api/src/main/java/org/apache/rya/api/domain/StatementMetadata.java]: Nope.  Good catch. You're right.

**Comment by pujav65** [comment]
asfbot build

**Pull request closed without merging**